### PR TITLE
Seal internal private types analyzer

### PR DESCRIPTION
**Pull request by NewellClark**
Fix [#49944](https://github.com/dotnet/runtime/issues/49944).
As per the [design meeting](https://github.com/dotnet/runtime/issues/49944#issuecomment-820636274), this analyzer ignores `InternalsVisibleToAttribute`s.
I've implemented a fixer, but due to [this issue](https://github.com/dotnet/roslyn/issues/51653) the code fixer won't show up in the IDE because it is reported at the end of compilation.

I've also found violations in dotnet/runtime ([violations in src/tasks](https://github.com/dotnet/runtime/pull/59935), [violations in libs](https://github.com/dotnet/runtime/pull/59941)).

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SealInternalTypes.Fixer.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            context.RegisterCodeFix(codeAction, context.Diagnostics);
+            return Task.CompletedTask;
+
+            //  Local functions
+
+            async Task<Document> SealClassAsync(CancellationToken token)
+            {
+                var editor = await DocumentEditor.CreateAsync(context.Document, token).ConfigureAwait(false);
+                var root = await context.Document.GetSyntaxRootAsync(token).ConfigureAwait(false);
+                var declaration = root.FindNode(context.Diagnostics[0].Location.SourceSpan);
```
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SealInternalTypes.Fixer.cs:32]:
```suggestion
                var declaration = root.FindNode(context.Span);
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SealInternalTypes.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        private static void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            var candidateTypes = PooledConcurrentSet<INamedTypeSymbol>.GetInstance(SymbolEqualityComparer.Default);
+            var baseTypes = PooledConcurrentSet<INamedTypeSymbol>.GetInstance(SymbolEqualityComparer.Default);
+
+            context.RegisterSymbolAction(context =>
+            {
+                var type = (INamedTypeSymbol)context.Symbol;
+
+                if (!IsExternallyVisible(type) && type.TypeKind is TypeKind.Class && !type.IsAbstract && !type.IsStatic && !type.IsSealed)
```
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SealInternalTypes.cs:46]: Keep cheap checks first.

```suggestion
                if (!type.IsAbstract && !type.IsStatic && !type.IsSealed && type.TypeKind is TypeKind.Class && !IsExternallyVisible(type))
```

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/SealInternalTypesTests.cs [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+using VerifyCS = Test.Utilities.CSharpCodeFixVerifier<
+    Microsoft.NetCore.Analyzers.Runtime.SealInternalTypes,
+    Microsoft.NetCore.Analyzers.Runtime.SealInternalTypesFixer>;
+using VerifyVB = Test.Utilities.VisualBasicCodeFixVerifier<
+    Microsoft.NetCore.Analyzers.Runtime.SealInternalTypes,
+    Microsoft.NetCore.Analyzers.Runtime.SealInternalTypesFixer>;
+
+namespace Microsoft.NetCore.Analyzers.Runtime.UnitTests
+{
+    public class SealInternalTypesTests
```
Comment (Youssef1313) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/SealInternalTypesTests.cs:16]: Can you tests for:

- Generic types (ie, a test that will fail if we remove `.OriginalDefinition` from the implementation)
- Partial classes (specifically for VB - sometimes `GetModifiers` API behavior is a bit unexpected depends on whether we pass the ClassBlock or ClassStatement node)

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SealInternalTypes.Fixer.cs (hunk 1 above)
Comment (NewellClark) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SealInternalTypes.Fixer.cs]: Fixed.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/SealInternalTypesTests.cs [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+                {
+                    Sources =
+                    {
+                        @"
+Friend Class Base
+End Class", @"
+Partial Friend NotInheritable Class Derived : Inherits Base
+End Class", @"
+Partial Friend NotInheritable Class Derived : Inherits Base
+End Class"
```
Comment (Youssef1313) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/SealInternalTypesTests.cs:366]: I'm concerned about adding attempting to seal a partial in a generated file. I'm not sure how other analyzers deal with similar scenarios. @Evangelink Have you got any idea?

cc @sharwell as well.

Comment (Youssef1313) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/SealInternalTypesTests.cs:366]: @NewellClark I think if one part of the symbol is in non-generated code, we'll get the symbol reported. Testing this should be easy, just add multiple sources, with one of them having `// <autogenerated>` at the top.

**Review by mavasani — APPROVED** [review]
Overall LGTM.

Note that the code fixer is never going to show up in VS as this is a compilation end diagnostic. It is still useful to retain it so a fix all can be applied from command line tooling, such as AnalyzerRunner or dotnet-format. In future, if we add some incremental analysis API such that compilation end analyzers and their code fixes can be enabled in the IDE, hopefully we will get the code fix to show up in VS.

Also calling out that compilation end diagnostics are essentially build-only diagnostics, which will only show up on build if the user bumps the severity to a warning or sets AnalysisMode to AllEnabledByDefault. They will only show up in live/intellisense diagnostics if full solution analysis mode is enabled. So basically, this will be disabled by default analyzer.

**Comment by mavasani** [comment]
@NewellClark Can you please resolve the merge conflicts?

**Comment by pranavkm** [comment]
@mavasani is this good to merge?

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/SealInternalTypesTests.cs (hunk 4 above)
Comment (Evangelink) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/SealInternalTypesTests.cs:366]: Sorry for the late answer! I don't know but I think adding the test should be easy and will document the behavior anyways.

**Comment by Evangelink** [comment]
It would be nice to add the test @Youssef1313 was mentioning regarding `partial` classes with some part auto-generated.

**Comment by mavasani** [comment]
@pranavkm We need to resolve merge conflicts and respond to review feedback from @Youssef1313 and @Evangelink. This should be very close to merge.

**Comment by pranavkm** [comment]
FYI @BrennanConroy

**Comment by meziantou** [comment]
@ericstj [ask me](https://github.com/dotnet/runtime/issues/49944#issuecomment-1075577520) to post my [remark here](https://github.com/dotnet/runtime/issues/49944#issuecomment-1073429272)

I think the analyzer should exclude classes declared as `[ComImport]`/`[CoClass]` from the analysis as the fix may break the compilation:

````c#
using System.Runtime.InteropServices;

_ = (IFileSaveDialog)new FileSaveDialogRCW(); // If FileSaveDialogRCW is sealed: error CS0030: Cannot convert type 'FileSaveDialogRCW' to 'IFileSaveDialog'

[ComImport]
[Guid("C0B4E2F3-BA21-4773-8DBA-335EC946EB8B")]
[InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
[CoClass(typeof(FileSaveDialogRCW))]
internal interface IFileSaveDialog
{
}

[ComImport]
[ClassInterface(ClassInterfaceType.None)]
[Guid("C0B4E2F3-BA21-4773-8DBA-335EC946EB8B")]
class FileSaveDialogRCW // 👈 sealed is not valid here because of `(IFileSaveDialog)new NativeFileSaveDialog()`
{
}
````

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx [hunk 5]
```diff
@@ -1795,6 +1795,18 @@
   <data name="DoNotAlwaysSkipTokenValidationInDelegatesMessage" xml:space="preserve">
     <value>The {0} is set to a function that is always returning true. By setting the validation delegate, you are overriding default validation and by always returning true, this validation is completely disabled.</value>
   </data>
+  <data name="SealInternalTypesDescription" xml:space="preserve">
+    <value>When a type is not accessible outside its assembly and has no subtypes within its containing assembly, it can be safely sealed. Sealing types can improve performance.</value>
+  </data>
+  <data name="SealInternalTypesMessage" xml:space="preserve">
+    <value>Type '{0}' can be sealed because it has no subtypes in its containing assembly and is not externally visible</value>
```
Comment (abatishchev) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:1802]: Some sentences are missing a dot at the end.

Comment (Youssef1313) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:1802]: @abatishchev Messages end with a period if and only if it's multiple statements. For a single-statement message, they shouldn't end with a period.

**Comment by stephentoub** [comment]
@NewellClark, I'd like to see this merged. Are you still able to work on it? If not, I'll take it over. Thanks for your work on it!

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SealInternalTypes.cs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+            context.RegisterSymbolAction(context =>
+            {
+                var type = (INamedTypeSymbol)context.Symbol;
+
+                if (type.TypeKind is TypeKind.Class &&
+                    !type.IsAbstract &&
+                    !type.IsStatic &&
+                    !type.IsSealed &&
+                    !type.IsExternallyVisible() &&
+                    (comImportAttributeType is null || !type.HasAttribute(comImportAttributeType)))
```
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SealInternalTypes.cs:53]: HasAttribute already does null check:

https://github.com/dotnet/roslyn-analyzers/blob/334bb06a340f279d0c6de7d2db54a5cefb0506e0/src/Utilities/Compiler/Extensions/ISymbolExtensions.cs#L642-L645

```suggestion
                    !type.HasAttribute(comImportAttributeType))
```

**Pull request merged**